[feat] Popup : #969 listview soql autofill

KRIMOD opened the pull request:
## Describe your changes
- When opening **Data Export** from a Salesforce List View, the extension now detects the current List View context (object + filterName) and calls the List View Describe REST API.
- The SOQL editor is automatically pre-populated with:
  - The fields that are displayed as columns in the List View.
  - The filters defined on the List View (translated into the `WHERE` clause where possible).
- If the List View context cannot be determined or the describe call fails, the behavior falls back to the existing default sample query so existing workflows are not impacted.
- A toggle feature on the settings has been added. 
- Updated `CHANGES.md` to document the new behavior.

## Issue ticket number and link
Implemented #969 – [Auto-populate SOQL editor from Salesforce List View context](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/969)

## Notes
Diff on popup.js showing a lot more changes than actually done because of an issue with the linter after running ` npx eslint --config eslint.config.mjs  addon/popup.js --fix`. If you have any idea on how to make a cleaner commit, I ll be happy to update it. 

## Checklist before requesting a review
- [x] I have read and understand the [Contributions section](https://github.com/tprouvot/Salesforce-Inspector-reloaded#contributions)
- [x] Target branch is releaseCandidate and not master
- [x] I have performed a self-review of my code
- [-] I ran the [unit tests](https://github.com/tprouvot/Salesforce-Inspector-reloaded#unit-tests) and my PR does not break any tests
  - Tests failing at the "Data Import" Section even without the implemented changes with the following error : `Error Message: "Cannot read properties of undefined (reading 'indexOf')"`  
- [x] I documented the changes I've made on the [CHANGES.md](https://github.com/tprouvot/Salesforce-Inspector-reloaded/blob/master/CHANGES.md) and followed actual conventions
- [-] I added a new section on [how-to.md](https://github.com/tprouvot/Salesforce-Inspector-reloaded/blob/master/docs/how-to.md) (optional)

KRIMOD committed:
v1 : added  query builder from listview on data export soql editor and enable listview feature flag on data export settings

KRIMOD committed:
doc : added entry to CHANGES.md

tprouvot commented on CHANGES.md:
```diff
@@ -17,6 +17,7 @@
 - `Show All Data` Support keyboard shortcut to save edited record values [feature 951](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/951) request by [Mohak Gaur](https://github.com/mohakgaurrr)
 - `Documentation` Revised instructions for creating an External Client App, reflecting the deprecation of Connected Apps and added detailed steps for OAuth configuration and known issues related to Incognito mode [bug 962](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/962) (contribution by [Thomas Malidin Delabriere](https://github.com/tmalidin33))
 - `Authentication` Implement OAuth 2.0 Web Server Flow with [Proof Key for Code Exchange (PKCE)](https://help.salesforce.com/s/articleView?id=xcloud.remoteaccess_pkce.htm&type=5) for orgs with API Access Control enabled - [feature 873](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/873) (contribution by [Mehdi Cherfaoui](https://github.com/mehdicherf))
+- `Data Export` Auto-populate SOQL editor from Salesforce List View context [feature 969](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/969) (request and contribution by [KRIMOD](https://github.com/KRIMOD))
```
```suggestion
- `Data Export` Auto-populate SOQL editor from Salesforce List View context [feature 969](https://github.com/tprouvot/Salesforce-Inspector-reloaded/issues/969) (request and contribution by [Temam Abdelkrim](https://github.com/KRIMOD))

```

tprouvot commented on addon/options.js:
```diff
@@ -171,6 +171,7 @@ class OptionsTabSelector extends React.Component {
           {option: Option, props: {type: "toggle", title: "Display Query Execution Time", key: "displayQueryPerformance", default: true}},
           {option: Option, props: {type: "toggle", title: "Show Local Time", key: "showLocalTime", default: false}},
           {option: Option, props: {type: "toggle", title: "Use SObject context on Data Export ", key: "useSObjectContextOnDataImpoltrink", default: true}},
+          {option: Option, props: {type: "toggle", title: "Enable List View Export", key: "enableListViewExport", default: true, tooltip: "Automatically populate Data Export query from current List View"}},
```
Since enabling this option consumes extra query, I would prefer to default it to false
```suggestion
          {option: Option, props: {type: "toggle", title: "Enable List View Export", key: "enableListViewExport", default: false, tooltip: "If enabled, Data Export link will be automatically populated with current ListView"}},
```

tprouvot commented on addon/popup.js:
```diff
@@ -185,11 +188,67 @@ class App extends React.PureComponent {
     });
   }
+  async getListViewQuery(url) {
+    if (localStorage.getItem("enableListViewExport") === "false") {
```
Since we changed the default option, the test also needs to be updated
```suggestion
    if (localStorage.getItem("enableListViewExport") !== "true") {
```

tprouvot commented on addon/popup.js:
```diff
@@ -185,11 +188,67 @@ class App extends React.PureComponent {
+    let listViewId = filterName;
+    let {sfHost} = this.props;
+
+    if (!filterName.startsWith("00B")) {
```
I couldn't find context where filterName contains the listview Id, did you ?

tprouvot commented on addon/popup.js:
```diff
@@ -185,11 +188,67 @@ class App extends React.PureComponent {
+
+    if (!filterName.startsWith("00B")) {
+      try {
+        const query = `SELECT Id FROM ListView WHERE SobjectType = '${sobjectName}' AND DeveloperName = '${filterName}'`;
```
Instead of having two queries, I would prefer to use composite API so that only one request is needed:

Here is what I tested (i was having different parameters btw)

``` js
async getListViewQuery(sobjectName, filterName) {
    if (localStorage.getItem("enableListViewExport") !== "true" || !sobjectName || !filterName) {
      return null;
    }

    try {
      if (!filterName.startsWith("00B")) {
        // Use Composite API to combine both requests into a single API call
        const query = `SELECT Id FROM ListView WHERE SobjectType = '${sobjectName}' AND DeveloperName = '${filterName}' LIMIT 1`;
        const compositePayload = {
          allOrNone: true,
          compositeRequest: [
            {
              method: "GET",
              url: `/services/data/v${apiVersion}/query/?q=${encodeURIComponent(query)}`,
              referenceId: "GetListViewInfo"
            },
            {
              method: "GET",
              url: `/services/data/v${apiVersion}/sobjects/${sobjectName}/listviews/@{GetListViewInfo.records[0].Id}/describe`,
              referenceId: "DescribeTheListView"
            }
          ]
        };

        const compositeRes = await sfConn.rest(
          `/services/data/v${apiVersion}/composite`,
          {method: "POST", body: compositePayload}
        );

        // Check if both requests succeeded
        if (compositeRes.compositeResponse
            && compositeRes.compositeResponse[1]?.httpStatusCode === 200) {
          const describeRes = compositeRes.compositeResponse[1].body;

          if (describeRes.query) {
            return describeRes.query;
          }
        }
      }
    } catch (e) {
      console.error(e);
    }
    return null;
  }
```

tprouvot commented on addon/popup.js:
```diff
@@ -170,6 +171,8 @@ class App extends React.PureComponent {
       }
       exportArg.set("query", query);
       importArg.set("sobject", e.contextSobject);
+    } else if (this.state.listViewQuery) {
```
Instead of using the state, I would prefer to keep the existing implementation and add a new property in `ensureKnownBrowserContext` method

``` js 
  ensureKnownBrowserContext() {
    let {contextUrl, onContextRecordChange} = this.props;
    if (contextUrl) {
      let recordId = getRecordId(contextUrl);
      let path = getSfPathFromUrl(contextUrl);
      let sobject = getSobject(contextUrl);
      let sobjectListview = getSobjectListview(contextUrl);
      let context = {
        contextRecordId: recordId,
        contextPath: path,
        contextSobject: sobject,
        contextSobjectListview: sobjectListview,
      };
      this.setState(context);
      onContextRecordChange(context);
    }
  }
  ```
  Decoupling the filtername extraction in 
``` js 
  function getSobjectListview(href) {
  const match = href
    ? href.match(/\/lightning\/o\/([^/]+)\/list\?filterName=([^&]+)/)
    : null;
  if (match) {
    return match[2]; // Return the filterName
  }
  return null;
}
  ``` 
  
And we would have in onContextRecordChange:
  
``` js 
      if (e.contextSobject && !e.contextSobjectListview && localStorage.getItem("useSObjectContextOnDataImportLink") !== "false") {
      let query = "SELECT Id FROM " + e.contextSobject;
      if (e.contextRecordId && (e.contextRecordId.length == 15 || e.contextRecordId.length == 18)) {
        query += " WHERE Id = '" + e.contextRecordId + "'";
      }
      exportArg.set("query", query);
      importArg.set("sobject", e.contextSobject);
    } else if (e.contextSobjectListview) {
      const listViewQuery = await this.getListViewQuery(e.contextSobject, e.contextSobjectListview);
      if (listViewQuery) {
        exportArg.set("query", listViewQuery);
      }
    }
```

tprouvot reviewed (commented):
Hi @KRIMOD 
Thanks for your PR ! It works well but I have few suggestions to address.
Another thing, the RecentlyViewed view does not work with current implementation but this is not the most used I guess for this feature

KRIMOD commented:
Hello @tprouvot, thank you for your feedback ! I ll look through everything you mentioned and get back to you.

KRIMOD committed:
Update CHANGES.md

Co-authored-by: Thomas Prouvot <35368290+tprouvot@users.noreply.github.com>

KRIMOD committed:
Update addon/options.js

dafault enableListViewExport to false

Co-authored-by: Thomas Prouvot <35368290+tprouvot@users.noreply.github.com>

KRIMOD committed:
Update addon/popup.js

Co-authored-by: Thomas Prouvot <35368290+tprouvot@users.noreply.github.com>

tprouvot commented:
Hi @KRIMOD 

I've seen some recent commits, should I test again ?

KRIMOD committed:
fix :
1. moved the state management for getListView logic
2. refactored the api calls to use composite api instead of 2 api calls

KRIMOD commented:
Hello @tprouvot, yes ! Please take a look when you have time. 

Concerning the filtering on the potential listview ID instead of api name, i was referencing this [document](https://www.fishofprey.com/2011/09/obscure-salesforce-object-key-prefixes.html). I am not entirely sure, but sometimes on some orgs, I get the listview id instead of the api name on the URL. My memory may be faulty, if it's not relevant I can get rid of it.

tprouvot committed:
Get rid of state for sobjectlist, open query in a new query tab when retrieved from query url parameter

tprouvot committed:
Merge branch 'releaseCandidate' into feat/969-listview-soql-autofill
